File: Problem2.js
// Time Complexity : O(n 2^n) n for checking Palindrome and 2^n for recursion
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes

var result;
var partition = function (s) {
    result = [];
    if (!s || s.length === 0) return result;
    helper(s, 0, []);
    return result;
};

var helper = function (s, index, path) {
    //base
    if (index === s.length) {
        result.push(path.slice());
        return;
    }

    //logic
    for (let i = index; i < s.length; i++) {
        if (isPalindrome(s, index, i)) {
            //action
            path.push(s.substring(index, i + 1));
            //recurse
            helper(s, i + 1, path);
            //bactrack
            path.pop();
        }
    }

}

var isPalindrome = function (s, l, r) {
    while (l < r) {
        if (s[l] !== s[r]) return false;
        l++;
        r--;
    }
    return true;
}

File: Problem2.cs
public class PalindromePartitioning
    {
        // Time Complexity : O(n * 2^n)
        // Space Complexity : O(n) - depth of the tree - length of the string
        // Did this code successfully run on Leetcode : Yes
        // Any problem you faced while coding this : No

        IList<IList<string>> result;
        public IList<IList<string>> Partition(string s)
        {
            result = new List<IList<string>>();
            helper(s, 0, new List<string>());
            return result;

        }

        private void helper(string s, int pivot, List<string> path)
        {
            //base case
            if (pivot == s.Length)
            {
                result.Add(new List<string>(path));
            }

            //logic
            for(int i = pivot; i < s.Length; i++)
            {
                string subStr = s.Substring(pivot, (i - pivot) + 1);
                //check if substring is palindrome
                if(isPalindrome(subStr))
                {
                    //action
                    path.Add(subStr);
                    //recurse
                    helper(s, i + 1, path);
                    //backtrack
                    path.RemoveAt(path.Count - 1);

                }
            }
        }

        private bool isPalindrome(string s)
        {
            int start = 0;
            int end = s.Length - 1;
            while(start < end)
            {
                if (s[start] != s[end])
                    return false;

                start++;
                end++;
            }
            return true;
        }
}
